src/iop/colorin.c

```diff
@@ -1764,6 +1764,89 @@ void reload_defaults(dt_iop_module_t *module)
     color_profile = DT_COLORSPACE_EMBEDDED_ICC;
   }
 
+  // We'll update the input profile hint with information on the embedded ICC profile if it exists.
+  // Since the tooltip info is now image dependent, we need to set the tooltip for each image change.
+
+  // We need gui_data to access widget in order to change tooltip.
+  dt_iop_colorin_gui_data_t *g = (dt_iop_colorin_gui_data_t *) module->gui_data;
+
+  // reload_defaults() can be called with unavailable (i.e., NULL) gui_data.
+  // In this case, we have nothing to do with tooltips.
+  if(!g) goto skip_tooltip_changing;
```

Comment:
Agreed!

Reply:
We all love Dijkstra, but his anti-goto crusade had a completely different context. And several decades of practical programming have shown that the article "Excess levels of code nesting considered harmful" should be written now.

The little old "goto" helps clean up code as was demonstrated repeatedly by the Linux kernel and C library. Linux kernel coding style guide lists the advantages of using goto:
- unconditional statements are easier to understand and follow
- nesting is reduced

For me, starting from the three levels of nested conditionals (especially multi-screen "long" if's) is hard to read and understand code that I try to avoid, such depth is a sign of poorly designed code. An extra level of code nesting is usually much worse for readability than a logical, sufficiently local and sufficiently commented goto.

At the end of writing the code, I just replaced the extra level of nesting with a goto bypass, because the code is clearer that way. Not sure I'd agree to make a change that I would perceive as degrading the code even for the sake of getting it accepted. :(

Reply:
Maybe an early return could be common middle ground in that _everybody_ hates _those_!

Moving all the GUI code to the end and then an `if (!g) return;` should work without having to hunt for the goto label.

Reply:
We already have " 'Goto considered harmful' considered harmful" from 1987.

Reply:
> Moving all the GUI code to the end and then an `if (!g) return;` should work without having to hunt for the goto label.

This is bad programming style because:
- `if (!g) return;` is less clear  then goto to well-named self-explanatory goto destination, like `if(!g) goto skip_tooltip_changing;`
- The correctness of such a code relies on the fact that the following code for working with tooltip will be at the very end of the function. If one day a programmer who makes changes to this function does not notice this non-obvious fact and adds unrelated code to the very end of the function, it will not work correctly. This is very fragile programming.

And we don't have to hunt for the goto label. Of course, when writing something incomprehensible like `goto LABEL1;` in the style of ancient BASIC or FORTRAN, you will have to look for this label and guess why we were sent here. It's just sloppily written and poorly commented code. But it is not needed when the self-explanatory name of the destination makes the goto command self-documenting.

Reply:
> We already have " 'Goto considered harmful' considered harmful" from 1987.

And let me quote the main points of this letter:


"__Although the argument was academic and unconvincing, its title seems to have become fixed in the mind of every programming manager and methodologist. Consequently, the notion that the GOTO is harmful is accepted almost universally, without question or doubt.__"

"__This has caused incalculable harm to the field of programming, which has lost an efficacious tool. It is like butchers banning knives because workers sometimes cut themselves. Programmers must devise eIaborate workarounds, use extra flags, nest statements excessively, or use gratuitous subroutines. The result is that GOTOless programs are harder and costlier to create, test, and modify. The cost to business has already been hundreds of millions of dollars in excess development and maintenance costs, plus the hidden cost of programs never developed due to insufficient resources.__"

"__All of my experiences compel me to conclude that it is time to part from the dogma of GOTO-less programming. It has failed to prove its merit.__"

Reply:
> This is bad programming style because:

Now if somebody from the other side could also come out against this idea, then we would have proven my hunch that

> an early return could be common middle ground in that everybody hates those!

and we could move forward with this PR!

(but maybe that's not the point anyway and we all just want a good argument. Cue Monty Python "This isn't an argument!")

Reply:
That little thing?  That's not an argument.   THIS is an argument!

(or am I mixing up movies again? :-) :-)

Reply:
Give that man a CIGAR!